rasa/architecture_prototype/graph_components.py

```diff
@@ -0,0 +1,233 @@
+        domain.persist(target_file)
+        return domain
+
+    def provide(self) -> Domain:
+        return self._domain
+
+
+class StoryGraphReader(ProjectReader):
+    def read(self, project: Text) -> StoryGraph:
+        importer = self.load_importer(project)
+
+        return rasa.utils.common.run_in_loop(importer.get_stories())
+
+
+class TrackerGenerator(GraphComponent):
+    def generate(
+        self, domain: Domain, story_graph: StoryGraph
+    ) -> List[TrackerWithCachedStates]:
+        generated_coroutine = rasa.core.training.load_data(story_graph, domain,)
+        return rasa.utils.common.run_in_loop(generated_coroutine)
+
+
+class StoryToTrainingDataConverter(GraphComponent):
+    def convert_for_training(self, story_graph: StoryGraph) -> TrainingData:
+        messages = []
+        for step in story_graph.story_steps:
+            messages += self._convert_tracker_to_messages(step.events)
+
+        # Workaround: add at least one end to end message to initialize
+        # the `CountVectorizer` for e2e. Alternatives: Store information or simply config
+        messages.append(
+            Message(
+                data=UserUttered(
+                    text="hi", use_text_for_featurization=True
+                ).as_sub_state()
+            )
+        )
+        return TrainingData(training_examples=messages)
+
+    def _convert_tracker_to_messages(self, events: List[Event]) -> List[Message]:
+        messages = []
+        for event in events:
+            if isinstance(event, ActionExecuted):
+                messages.append(Message(data=event.as_sub_state()))
+
+            if isinstance(event, UserUttered):
+                if event.use_text_for_featurization is None:
+                    event.use_text_for_featurization = False
+                    messages.append(Message(data=event.as_sub_state()))
+
+                    event.use_text_for_featurization = True
+                    messages.append(Message(data=event.as_sub_state()))
+
+                    event.use_text_for_featurization = None
+                else:
+                    messages.append(Message(data=event.as_sub_state()))
+
+        return messages
+
+    def convert_for_inference(self, tracker: DialogueStateTracker) -> List[Message]:
+        return self._convert_tracker_to_messages(tracker.events)
+
+
+class MessageToE2EFeatureConverter(GraphComponent):
+    """Collects featurised messages for use by an e2e policy."""
+    def convert(
+        self, messages: Union[TrainingData, List[Message]]
+    ) -> Dict[Text, Message]:
+        if isinstance(messages, TrainingData):
+            messages = messages.training_examples
+        additional_features = {}
+        for message in messages:
+            key = next(
+                k
+                for k in message.data.keys()
+                if k in {ACTION_NAME, ACTION_TEXT, INTENT, TEXT}
+            )
+            additional_features[key] = message
```

Comment:
Was just cherry picking some parts to get the new featurization working without the graph and noticed that this code is wrong. Should probably be something like:
```
        for message in messages:
            key = next(
                v
                for k, v in message.data.items()
                if k in {ACTION_NAME, ACTION_TEXT, INTENT, TEXT}
            )
            additional_features[key] = message
```
otherwise you are literally saving `additional_features["action_name"] = message` for almost every message

Reply:
oh dear - yes, it was very late when I did this 🤦🏻

Reply:
thanks for the catch!